pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -823,9 +823,11 @@ func (p *PriorityQueue) Pop() (*framework.QueuedPodInfo, error) {
 		}
 	}
 
+	// Update metrics and reset the set of unschedulable plugins for the next attempt.
 	for plugin := range pInfo.UnschedulablePlugins {
 		metrics.UnschedulableReason(plugin, pInfo.Pod.Spec.SchedulerName).Dec()
 	}
+	pInfo.UnschedulablePlugins.Clear()
```

Comment:
LGTM. Can we make it UTed?

Reply:
What does "UTed" mean?

Reply:
Probably "unit tested" - I don't know the unit tests for this code well, but I can have a look where such a test might fit.

Reply:
yup, unit tested. It'd be nice to cover the code change in UT, or integration test if you think that's easier.

Reply:
I couldn't find any unit tests for the ["unschedulable_pods" metric](https://github.com/kubernetes/kubernetes/blob/cd04f47e8b287068e8bf0bfae6596b2606d54929/pkg/scheduler/metrics/metrics.go#L200-L206), which is the other reason why `Pop` needs `pInfo.UnschedulablePlugins`.

Looking through the code I am not convinced that this metric was handled correctly at all.

For example, when a pod gets added first to the unschedulable queue and then moved to the backoff queue, should it still count as "unschedulable"? I don't think so. Not only does the code not decrement in this case, it *increments* again, counting the same pod twice, because `AddUnschedulableIfNotPresent` gets called twice - or at least that is what happens when the `addUnschedulablePodBackToUnschedulablePods` and `moveAllToActiveOrBackoffQ` operations get called.

I don't have time to untangle this right now. Can we merge this PR as-is as a hotfix and then clean up later? "unschedulable_pods"  is alpha, so it's okayish that remains broken in 1.28.

I created https://github.com/kubernetes/kubernetes/issues/120395 with instructions and a preliminary unit test.

Reply:
An integration test is harder if we want it to be reliable. We could enable the [`SchedulingWithMultipleResourceClaims`](https://github.com/kubernetes/kubernetes/blob/294bde0079a0d56099cf8b8cf558e3ae7230de12/test/integration/scheduler_perf/config/performance-config.yaml#L785C9-L785C45) test so that it runs as part of `pull-kubernetes-integration`, but it was only failing occasionally.

I'm also not sure how I feel as SIG-Testing tech lead about enabling a test that depends on an alpha feature in jobs for stable features. It gets done for unit tests because there really is no good alternative. It's not much better for integration testing, though, there's also no periodic "make alpha-integration-test".

I guess it's okay and could get reverted if it really causes problems.

Reply:
Let's forget about metrics in this PR (thanks for opening the follow up issue).

Still, this change needs a test, unit or integration, that proves that clearing the unschedulable plugins is enough for this pod to make it to backoff instead of unschedulable.

Also, I don't think this is exclusive to DRA. For example, this might reproduce with any plugin that returns unschedulable, followed by a retry where the pod fits but fails in Bind.

Reply:
Another way to express this bug is that a scheduling attempt that is an Error, as opposed to Unschedulable(AndUnresolvable), should put the Pod into backoff in all cases.

@sanposhiho weren't you working on a PR along these lines?

Reply:
> Still, this change needs a test, unit or integration, that proves that clearing the unschedulable plugins is enough for this pod to make it to backoff instead of unschedulable.

You are right of course. It seemed to help with my stress testing, but the error was a rare concurrent update conflict and perhaps I was simply not hitting that timing condition anymore.

I wonder how the backoff queue was chosen in the first place. My fix might not be complete yet.

`handleSchedulingFailure` just [calls `AddUnschedulableIfNotPresent`](https://github.com/kubernetes/kubernetes/blob/04292dd94bd9999dfced146f4de8b81a0e466e06/pkg/scheduler/schedule_one.go#L1021C36-L1021C64), presumably with an empty `UnschedulablePlugins`. There's no explicit "requeue because of failure", is there?

`AddUnschedulableIfNotPresent` then calls `determineSchedulingHintForInFlightPod` which [returns `QueueSkip`](https://github.com/kubernetes/kubernetes/blob/04292dd94bd9999dfced146f4de8b81a0e466e06/pkg/scheduler/internal/queue/scheduling_queue.go#L636-L648) (no unschedulable plugins, no concurrent events). The code before my recent concurrent event tracking PR [did the same ](https://github.com/kubernetes/kubernetes/blob/04292dd94bd9999dfced146f4de8b81a0e466e06/pkg/scheduler/internal/queue/scheduling_queue.go#L636-L648).

`requeuePodViaQueueingHint` then [uses the unschedulable queue](https://github.com/kubernetes/kubernetes/blob/04292dd94bd9999dfced146f4de8b81a0e466e06/pkg/scheduler/internal/queue/scheduling_queue.go#L1057-L1061). It never reaches the "must pod back off?" part.

It's not better without hints. `addUnschedulableWithoutQueueingHint` only ever uses the backoff queue if there was [a concurrent event](https://github.com/kubernetes/kubernetes/blob/04292dd94bd9999dfced146f4de8b81a0e466e06/pkg/scheduler/internal/queue/scheduling_queue.go#L695-L700).

I'll push unit tests that I added, but both are failing right now and I don't know how to fix that.

Reply:
> handleSchedulingFailure just [calls AddUnschedulableIfNotPresent](https://github.com/kubernetes/kubernetes/blob/04292dd94bd9999dfced146f4de8b81a0e466e06/pkg/scheduler/schedule_one.go#L1021C36-L1021C64), presumably with an empty UnschedulablePlugins. There's no explicit "requeue because of failure", is there?

There isn't, but I think that's what we should fix.

Maybe no unschedulable plugins is enough signal that this was an error? But I would prefer if we make it more explicit.

Reply:
I just verified that an error during Bind would still cause the Pod to go into Backoff, as proven by this test https://github.com/kubernetes/kubernetes/pull/120472 (it was flaky, but mostly succeeding).

The reason is that when we fail during the binding cycle, we call `ForgetPod`, which produces a fake event for "deleted scheduled pod". This event is enough to consistently put the Pod that doesn't have any unschedulable plugins into Backoff.

But we don't run ForgetPod if Reserve fails, which I think is what happens in DRA.

Reply:
>  Another way to express this bug is that a scheduling attempt that is an Error, as opposed to Unschedulable(AndUnresolvable), should put the Pod into backoff in all cases.

> Maybe no unschedulable plugins is enough signal that this was an error?

We have the PR for this already. It'll be merged once rebasing done.

https://github.com/kubernetes/kubernetes/pull/119105

Reply:
> But we don't run ForgetPod if Reserve fails, which I think is what happens in DRA.

We actually do I think - when any Reserve plugin returns non-success code:

https://github.com/kubernetes/kubernetes/blob/130a5a423fb1b9b6d317358ba3b6cf24f586c075/pkg/scheduler/schedule_one.go#L201-L207

Reply:
> But we don't run ForgetPod if Reserve fails, which I think is what happens in DRA.

Then this PR maybe actually fixes the issue? This PR is semantically correct regardless, so we should merge it. But maybe we can focus the UT on having the plugins cleared or have a UT at a higher level (maybe scheduler_test.go)

Reply:
I agree, it's hard to be certain about "correct" behavior at this level because the unit tests must faithfully replicate what happens in "real" scheduling attempts and that seems to be ambiguous.

Let me remove the unit tests and archive them here instead:

<details>

```patch
commit 2ce67afa02be3912576604900ea9eb7eefdb393d
Author: Patrick Ohly <patrick.ohly@intel.com>
Date:   Wed Sep 6 10:52:38 2023 +0200

    WIP: unit tests

diff --git a/pkg/scheduler/internal/queue/scheduling_queue_test.go b/pkg/scheduler/internal/queue/scheduling_queue_test.go
index 12b92bb7bdd..24c0b72d05a 100644
--- a/pkg/scheduler/internal/queue/scheduling_queue_test.go
+++ b/pkg/scheduler/internal/queue/scheduling_queue_test.go
@@ -542,6 +542,84 @@ func Test_InFlightPods(t *testing.T) {
 				},
 			},
 		},
+		{
+			name:                         "failed pod is placed in backoff queue with hints",
+			isSchedulingQueueHintEnabled: true,
+			initialPods:                  []*v1.Pod{pod},
+			actions: []action{
+				{callback: func(t *testing.T, q *PriorityQueue) { poppedPod = popPod(t, q, pod) }},
+				{callback: func(t *testing.T, q *PriorityQueue) {
+					logger, _ := ktesting.NewTestContext(t)
+					// Unschedulable.
+					poppedPod.UnschedulablePlugins = sets.New("fooPlugin1")
+					if err := q.AddUnschedulableIfNotPresent(logger, poppedPod, q.SchedulingCycle()); err != nil {
+						t.Errorf("Unexpected error from AddUnschedulableIfNotPresent: %v", err)
+					}
+				}},
+				{eventHappens: &PvAdd}, // Active again.
+				{callback: func(t *testing.T, q *PriorityQueue) { poppedPod = popPod(t, q, pod) }},
+				{callback: func(t *testing.T, q *PriorityQueue) {
+					logger, _ := ktesting.NewTestContext(t)
+					// Failed (i.e. no UnschedulablePlugins). Should go to backoff.
+					if err := q.AddUnschedulableIfNotPresent(logger, poppedPod, q.SchedulingCycle()); err != nil {
+						t.Errorf("Unexpected error from AddUnschedulableIfNotPresent: %v", err)
+					}
+				}},
+			},
+			wantBackoffQPodNames: []string{pod.Name},
+			wantInFlightPods:     nil,
+			wantInFlightEvents:   nil,
+			queueingHintMap: QueueingHintMapPerProfile{
+				"": {
+					PvAdd: {
+						{
+							// The hint fn tells that this event makes a Pod scheudlable immediately.
+							PluginName:     "fooPlugin1",
+							QueueingHintFn: queueHintReturnQueueImmediately,
+						},
+					},
+				},
+			},
+		},
+		{
+			name:                         "failed pod is placed in backoff queue without hints",
+			isSchedulingQueueHintEnabled: false,
+			initialPods:                  []*v1.Pod{pod},
+			actions: []action{
+				{callback: func(t *testing.T, q *PriorityQueue) { poppedPod = popPod(t, q, pod) }},
+				{callback: func(t *testing.T, q *PriorityQueue) {
+					logger, _ := ktesting.NewTestContext(t)
+					// Unschedulable.
+					poppedPod.UnschedulablePlugins = sets.New("fooPlugin1")
+					if err := q.AddUnschedulableIfNotPresent(logger, poppedPod, q.SchedulingCycle()); err != nil {
+						t.Errorf("Unexpected error from AddUnschedulableIfNotPresent: %v", err)
+					}
+				}},
+				{eventHappens: &PvAdd}, // Active again.
+				{callback: func(t *testing.T, q *PriorityQueue) { poppedPod = popPod(t, q, pod) }},
+				{callback: func(t *testing.T, q *PriorityQueue) {
+					logger, _ := ktesting.NewTestContext(t)
+					// Failed (i.e. no UnschedulablePlugins). Should go to backoff.
+					if err := q.AddUnschedulableIfNotPresent(logger, poppedPod, q.SchedulingCycle()); err != nil {
+						t.Errorf("Unexpected error from AddUnschedulableIfNotPresent: %v", err)
+					}
+				}},
+			},
+			wantBackoffQPodNames: []string{pod.Name},
+			wantInFlightPods:     nil,
+			wantInFlightEvents:   nil,
+			queueingHintMap: QueueingHintMapPerProfile{
+				"": {
+					PvAdd: {
+						{
+							// The hint fn tells that this event makes a Pod scheudlable immediately.
+							PluginName:     "fooPlugin1",
+							QueueingHintFn: queueHintReturnQueueImmediately,
+						},
+					},
+				},
+			},
+		},
 	}
 
 	for _, test := range tests {
```

</details>

Then I'll let you decide whether you want to merge this PR and add unit tests elsewhere, perhaps in https://github.com/kubernetes/kubernetes/pull/119105.

Reply:
I would still leave a test that makes sure the list of unschedulable plugins in a second attempt.

Reply:
Done. The unit test fails (as it should) without the new `pInfo.UnschedulablePlugins.Clear()` in `Pop`.